CI: Travis: ClusterMeshServicesTestSuite.TestClusterMeshServicesUpdate: obtained k8s.CacheAction = 0 ("service-updated")

## CI failure

https://app.travis-ci.com/github/cilium/cilium/jobs/585481519#L527

```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃   FAIL  package: github.com/cilium/cilium/pkg/clustermesh   ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
--- FAIL: Test (90.63s)
                       
level=info msg="Creating etcd client" ConfigPath= KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Initializing identity allocator" subsys=identity-cache
level=info msg="Allocating identities between range" cluster-id=0 max=393215 min=327680 subsys=identity-cache
level=info msg="New remote cluster configuration" clusterName=cluster1 config=/tmp/multicluster1343203379/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="New remote cluster configuration" clusterName=cluster2 config=/tmp/multicluster1343203379/cluster2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Got lease ID 49ec83ce3868e3ea and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e3ec" subsys=kvstore
level=info msg="Initial etcd session established" config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster1343203379/cluster2 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster1343203379/cluster2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster1343203379/cluster1 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster1343203379/cluster1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config= endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Got lease ID 49ec83ce3868e3ee and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e3f7" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster1343203379/cluster1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Got lease ID 49ec83ce3868e3f4 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e3f9" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster1343203379/cluster2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster1343203379/cluster2 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=cluster2 config=/tmp/multicluster1343203379/cluster2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster1343203379/cluster1 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=cluster1 config=/tmp/multicluster1343203379/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=cluster2 config=/tmp/multicluster1343203379/cluster2 kvstoreErr="<nil>" kvstoreStatus="etcd: 1/1 connected, lease-ID=49ec83ce3868e3f4, lock lease-ID=49ec83ce3868e3f9, has-quorum=true: http://127.0.0.1:4002 - 3.3.20 (Leader)" subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=cluster1 config=/tmp/multicluster1343203379/cluster1 kvstoreErr="<nil>" kvstoreStatus="etcd: 1/1 connected, lease-ID=49ec83ce3868e3ee, lock lease-ID=49ec83ce3868e3f7, has-quorum=true: http://127.0.0.1:4002 - 3.3.20 (Leader)" subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=cluster2 config=/tmp/multicluster1343203379/cluster2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=cluster2 config=/tmp/multicluster1343203379/cluster2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=cluster2 config=/tmp/multicluster1343203379/cluster2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=cluster1 config=/tmp/multicluster1343203379/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=cluster1 config=/tmp/multicluster1343203379/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=cluster1 config=/tmp/multicluster1343203379/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Initializing identity allocator" subsys=identity-cache
level=info msg="Allocating identities between range" cluster-id=0 max=393215 min=327680 subsys=identity-cache
level=info msg="New remote cluster configuration" clusterName=cluster1 config=/tmp/multicluster2779721787/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="New remote cluster configuration" clusterName=cluster2 config=/tmp/multicluster2779721787/cluster2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=cluster1 config=/tmp/multicluster2779721787/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="New remote cluster configuration" clusterName=cluster3 config=/tmp/multicluster2779721787/cluster3 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=cluster3 config=/tmp/multicluster2779721787/cluster3 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="New remote cluster configuration" clusterName=cluster1 config=/tmp/multicluster2779721787/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=cluster1 config=/tmp/multicluster2779721787/cluster1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=cluster2 config=/tmp/multicluster2779721787/cluster2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Creating etcd client" ConfigPath= KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Initializing identity allocator" subsys=identity-cache
level=info msg="Allocating identities between range" cluster-id=0 max=393215 min=327680 subsys=identity-cache
level=info msg="Got lease ID 49ec83ce3868e415 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e418" subsys=kvstore
level=info msg="Initial etcd session established" config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config= endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="New remote cluster configuration" clusterName=anFTOkXjdBIR1 config=/tmp/multicluster2484135826/anFTOkXjdBIR1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="New remote cluster configuration" clusterName=anFTOkXjdBIR2 config=/tmp/multicluster2484135826/anFTOkXjdBIR2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster2484135826/anFTOkXjdBIR1 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster2484135826/anFTOkXjdBIR2 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster2484135826/anFTOkXjdBIR2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster2484135826/anFTOkXjdBIR1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Got lease ID 49ec83ce3868e421 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e425" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster2484135826/anFTOkXjdBIR2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Got lease ID 49ec83ce3868e423 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e427" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster2484135826/anFTOkXjdBIR1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster2484135826/anFTOkXjdBIR2 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=anFTOkXjdBIR2 config=/tmp/multicluster2484135826/anFTOkXjdBIR2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster2484135826/anFTOkXjdBIR1 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=anFTOkXjdBIR1 config=/tmp/multicluster2484135826/anFTOkXjdBIR1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=anFTOkXjdBIR2 config=/tmp/multicluster2484135826/anFTOkXjdBIR2 kvstoreErr="<nil>" kvstoreStatus="Waiting for initial connection to be established" subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=anFTOkXjdBIR1 config=/tmp/multicluster2484135826/anFTOkXjdBIR1 kvstoreErr="<nil>" kvstoreStatus="Waiting for initial connection to be established" subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=anFTOkXjdBIR1 config=/tmp/multicluster2484135826/anFTOkXjdBIR1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=anFTOkXjdBIR1 config=/tmp/multicluster2484135826/anFTOkXjdBIR1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=anFTOkXjdBIR2 config=/tmp/multicluster2484135826/anFTOkXjdBIR2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=anFTOkXjdBIR2 config=/tmp/multicluster2484135826/anFTOkXjdBIR2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=anFTOkXjdBIR1 config=/tmp/multicluster2484135826/anFTOkXjdBIR1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=anFTOkXjdBIR2 config=/tmp/multicluster2484135826/anFTOkXjdBIR2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Creating etcd client" ConfigPath= KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Initializing identity allocator" subsys=identity-cache
level=info msg="Allocating identities between range" cluster-id=0 max=393215 min=327680 subsys=identity-cache
level=info msg="Got lease ID 49ec83ce3868e566 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e569" subsys=kvstore
level=info msg="Initial etcd session established" config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config= endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="New remote cluster configuration" clusterName=QOWlFtCVXost1 config=/tmp/multicluster4084475411/QOWlFtCVXost1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="New remote cluster configuration" clusterName=QOWlFtCVXost2 config=/tmp/multicluster4084475411/QOWlFtCVXost2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster4084475411/QOWlFtCVXost1 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster4084475411/QOWlFtCVXost1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster4084475411/QOWlFtCVXost2 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster4084475411/QOWlFtCVXost2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Got lease ID 49ec83ce3868e572 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e576" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster4084475411/QOWlFtCVXost1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Got lease ID 49ec83ce3868e574 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e578" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster4084475411/QOWlFtCVXost2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster4084475411/QOWlFtCVXost1 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=QOWlFtCVXost1 config=/tmp/multicluster4084475411/QOWlFtCVXost1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=QOWlFtCVXost1 config=/tmp/multicluster4084475411/QOWlFtCVXost1 kvstoreErr="<nil>" kvstoreStatus="etcd: 1/1 connected, lease-ID=49ec83ce3868e572, lock lease-ID=49ec83ce3868e576, has-quorum=true: http://127.0.0.1:4002 - 3.3.20 (Leader)" subsys=clustermesh
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster4084475411/QOWlFtCVXost2 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=QOWlFtCVXost2 config=/tmp/multicluster4084475411/QOWlFtCVXost2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=QOWlFtCVXost2 config=/tmp/multicluster4084475411/QOWlFtCVXost2 kvstoreErr="<nil>" kvstoreStatus="etcd: 1/1 connected, lease-ID=49ec83ce3868e574, lock lease-ID=49ec83ce3868e578, has-quorum=true: http://127.0.0.1:4002 - 3.3.20 (Leader)" subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=QOWlFtCVXost1 config=/tmp/multicluster4084475411/QOWlFtCVXost1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=QOWlFtCVXost1 config=/tmp/multicluster4084475411/QOWlFtCVXost1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=QOWlFtCVXost2 config=/tmp/multicluster4084475411/QOWlFtCVXost2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=QOWlFtCVXost2 config=/tmp/multicluster4084475411/QOWlFtCVXost2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=QOWlFtCVXost1 config=/tmp/multicluster4084475411/QOWlFtCVXost1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=QOWlFtCVXost2 config=/tmp/multicluster4084475411/QOWlFtCVXost2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Creating etcd client" ConfigPath= KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Initializing identity allocator" subsys=identity-cache
level=info msg="Allocating identities between range" cluster-id=0 max=393215 min=327680 subsys=identity-cache
level=info msg="Got lease ID 49ec83ce3868e583 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e586" subsys=kvstore
level=info msg="Initial etcd session established" config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config= endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="New remote cluster configuration" clusterName=QtJjmgqPdiTY1 config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="New remote cluster configuration" clusterName=QtJjmgqPdiTY2 config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster3651564398/QtJjmgqPdiTY2 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Got lease ID 49ec83ce3868e58f and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e591" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Creating etcd client" ConfigPath=/tmp/multicluster3651564398/QtJjmgqPdiTY1 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
level=info msg="Connecting to etcd server..." config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=QtJjmgqPdiTY2 config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Got lease ID 49ec83ce3868e593 and the session TTL is 0s" subsys=kvstore
level=info msg="Got lock lease ID 49ec83ce3868e595" subsys=kvstore
level=info msg="Initial etcd session established" config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 endpoints="[http://127.0.0.1:4002]" subsys=kvstore
level=info msg="Successfully verified version of etcd endpoint" config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20
level=info msg="Connection to remote cluster established" clusterName=QtJjmgqPdiTY1 config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=QtJjmgqPdiTY2 config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 kvstoreErr="<nil>" kvstoreStatus="etcd: 1/1 connected, lease-ID=49ec83ce3868e58f, lock lease-ID=49ec83ce3868e591, has-quorum=true: http://127.0.0.1:4002 - 3.3.20 (Leader)" subsys=clustermesh
level=info msg="Established connection to remote etcd" clusterName=QtJjmgqPdiTY1 config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 kvstoreErr="<nil>" kvstoreStatus="etcd: 1/1 connected, lease-ID=49ec83ce3868e593, lock lease-ID=49ec83ce3868e595, has-quorum=true: http://127.0.0.1:4002 - 3.3.20 (Leader)" subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=QtJjmgqPdiTY1 config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=QtJjmgqPdiTY1 config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=QtJjmgqPdiTY1 config=/tmp/multicluster3651564398/QtJjmgqPdiTY1 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="All resources of remote cluster cleaned up" clusterName=QtJjmgqPdiTY2 config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Remote cluster disconnected" clusterName=QtJjmgqPdiTY2 config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
level=info msg="Closing connection to remote etcd" clusterName=QtJjmgqPdiTY2 config=/tmp/multicluster3651564398/QtJjmgqPdiTY2 kvstoreErr="<nil>" kvstoreStatus= subsys=clustermesh
----------------------------------------------------------------------
FAIL: services_test.go:208: ClusterMeshServicesTestSuite.TestClusterMeshServicesUpdate
services_test.go:263:
    s.expectEvent(c, k8s.DeleteService, svcID, nil)
services_test.go:120:
    c.Assert(event.Action, Equals, action)
... obtained k8s.CacheAction = 0 ("service-updated")
... expected k8s.CacheAction = 1 ("service-deleted")
OOPS: 5 passed, 1 FAILED
```

--- a/operator/watchers/k8s_service_sync.go
+++ b/operator/watchers/k8s_service_sync.go
@@ -114,7 +114,7 @@ func StartSynchronizingServices(ctx context.Context, clientset k8sClient.Clients
 		store, err := store.JoinSharedStore(store.Configuration{
 			Prefix:                  serviceStore.ServiceStorePrefix,
 			SynchronizationInterval: 5 * time.Minute,
-			SharedKeyDeleteDelay:    0,
+			SharedKeyDeleteDelay:    func() *time.Duration { a := time.Duration(0); return &a }(),
 			KeyCreator: func() store.Key {
 				return &serviceStore.ClusterService{}
 			},

--- a/pkg/clustermesh/clustermesh.go
+++ b/pkg/clustermesh/clustermesh.go
@@ -6,6 +6,7 @@ package clustermesh
 import (
 	"context"
 	"fmt"
+	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
 
@@ -62,6 +63,14 @@ type Configuration struct {
 	RemoteIdentityWatcher RemoteIdentityWatcher
 
 	IPCache *ipcache.IPCache
+
+	// ServicesSharedKeyDeleteDelay is the delay before a shared service delete
+	// event is handled. This parameter is optional.
+	ServicesSharedKeyDeleteDelay *time.Duration
+
+	// NodesSharedKeyDeleteDelay is the delay before a shared node delete event
+	// is handled. This parameter is optional.
+	NodesSharedKeyDeleteDelay *time.Duration
 }
 
 // RemoteIdentityWatcher is any type which provides identities that have been

--- a/pkg/clustermesh/remote_cluster.go
+++ b/pkg/clustermesh/remote_cluster.go
@@ -180,6 +180,7 @@ func (rc *remoteCluster) restartRemoteConnection(allocator RemoteIdentityWatcher
 				remoteNodes, err := store.JoinSharedStore(store.Configuration{
 					Prefix:                  path.Join(nodeStore.NodeStorePrefix, rc.name),
 					KeyCreator:              rc.mesh.conf.NodeKeyCreator,
+					SharedKeyDeleteDelay:    rc.mesh.conf.NodesSharedKeyDeleteDelay,
 					SynchronizationInterval: time.Minute,
 					Backend:                 backend,
 					Observer:                rc.mesh.conf.NodeObserver(),
@@ -190,7 +191,8 @@ func (rc *remoteCluster) restartRemoteConnection(allocator RemoteIdentityWatcher
 				}
 
 				remoteServices, err := store.JoinSharedStore(store.Configuration{
-					Prefix: path.Join(serviceStore.ServiceStorePrefix, rc.name),
+					Prefix:               path.Join(serviceStore.ServiceStorePrefix, rc.name),
+					SharedKeyDeleteDelay: rc.mesh.conf.ServicesSharedKeyDeleteDelay,
 					KeyCreator: func() store.Key {
 						svc := serviceStore.ClusterService{}
 						return &svc

--- a/pkg/clustermesh/services_test.go
+++ b/pkg/clustermesh/services_test.go
@@ -82,13 +82,14 @@ func (s *ClusterMeshServicesTestSuite) SetUpTest(c *C) {
 	})
 	defer ipc.Shutdown()
 	cm, err := NewClusterMesh(Configuration{
-		Name:                  "test2",
-		ConfigDirectory:       dir,
-		NodeKeyCreator:        testNodeCreator,
-		nodeObserver:          &testObserver{},
-		ServiceMerger:         &s.svcCache,
-		RemoteIdentityWatcher: mgr,
-		IPCache:               ipc,
+		Name:                         "test2",
+		ConfigDirectory:              dir,
+		NodeKeyCreator:               testNodeCreator,
+		nodeObserver:                 &testObserver{},
+		ServiceMerger:                &s.svcCache,
+		RemoteIdentityWatcher:        mgr,
+		IPCache:                      ipc,
+		ServicesSharedKeyDeleteDelay: func() *time.Duration { a := time.Duration(0); return &a }(),
 	})
 	c.Assert(err, IsNil)
 	c.Assert(cm, Not(IsNil))
@@ -247,26 +248,35 @@ func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesUpdate(c *C) {
 	kvstore.Client().Set(context.TODO(), k, []byte(v))
 	s.expectEvent(c, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
 		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")] != nil
+			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")].Ports["http"].DeepEqual(
+				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080)) &&
+			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")] != nil &&
+			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")].Ports["http2"].DeepEqual(
+				loadbalancer.NewL4Addr(loadbalancer.TCP, 90))
 	})
 
 	k, v = s.prepareServiceUpdate("2", "90.0.185.196", "http", "8080")
 	kvstore.Client().Set(context.TODO(), k, []byte(v))
 	s.expectEvent(c, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
 		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")] != nil
+			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")].Ports["http"].DeepEqual(
+				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080)) &&
+			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")] != nil &&
+			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")].Ports["http"].DeepEqual(
+				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080))
 	})
 
 	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"1")
-	// The observer will have a defaults.NodeDeleteDelay time before it receives
-	// the event. For this reason we will trigger the delete events sequentially
-	// and only do the assertion in the end. This way we wait 30seconds for the
-	// test to complete instead of 30+30 seconds.
-	time.Sleep(2 * time.Second)
-	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"2")
+	s.expectEvent(c, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
+		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")] != nil &&
+			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")].Ports["http"].DeepEqual(
+				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080))
+	})
 
-	s.expectEvent(c, k8s.UpdateService, svcID, nil)
-	s.expectEvent(c, k8s.DeleteService, svcID, nil)
+	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"2")
+	s.expectEvent(c, k8s.DeleteService, svcID, func(event k8s.ServiceEvent) bool {
+		return len(event.Endpoints.Backends) == 0
+	})
 
 	swgSvcs.Stop()
 	c.Assert(testutils.WaitUntil(func() bool {

--- a/pkg/kvstore/store/store.go
+++ b/pkg/kvstore/store/store.go
@@ -54,9 +54,9 @@ type Configuration struct {
 	// are synchronized with the kvstore. This parameter is optional.
 	SynchronizationInterval time.Duration
 
-	// SharedKeyDeleteDelay is the delay before an shared key delete is
+	// SharedKeyDeleteDelay is the delay before a shared key delete is
 	// handled. This parameter is optional
-	SharedKeyDeleteDelay time.Duration
+	SharedKeyDeleteDelay *time.Duration
 
 	// KeyCreator is called to allocate a Key instance when a new shared
 	// key is discovered. This parameter is required.
@@ -87,8 +87,8 @@ func (c *Configuration) validate() error {
 		c.SynchronizationInterval = option.Config.KVstorePeriodicSync
 	}
 
-	if c.SharedKeyDeleteDelay == 0 {
-		c.SharedKeyDeleteDelay = defaults.NodeDeleteDelay
+	if c.SharedKeyDeleteDelay == nil {
+		c.SharedKeyDeleteDelay = func() *time.Duration { a := defaults.NodeDeleteDelay; return &a }()
 	}
 
 	if c.Backend == nil {
@@ -416,12 +416,16 @@ func (s *SharedStore) deleteSharedKey(name string) {
 
 	if ok {
 		go func() {
-			time.Sleep(s.conf.SharedKeyDeleteDelay)
+			var timeWindow time.Duration
+			if s.conf.SharedKeyDeleteDelay != nil {
+				timeWindow = *s.conf.SharedKeyDeleteDelay
+				time.Sleep(timeWindow)
+			}
 			s.mutex.RLock()
 			_, ok := s.sharedKeys[name]
 			s.mutex.RUnlock()
 			if ok {
-				s.getLogger().WithFields(logrus.Fields{"key": name, "timeWindow": s.conf.SharedKeyDeleteDelay}).
+				s.getLogger().WithFields(logrus.Fields{"key": name, "timeWindow": timeWindow}).
 					Warning("Received delete event for key which re-appeared within delay time window")
 				return
 			}

--- a/pkg/kvstore/store/store_test.go
+++ b/pkg/kvstore/store/store_test.go
@@ -171,7 +171,7 @@ func (s *StoreSuite) TestStoreOperations(c *C) {
 		Prefix:               rand.RandomString(),
 		KeyCreator:           newTestType,
 		Observer:             &observer{},
-		SharedKeyDeleteDelay: sharedKeyDeleteDelay,
+		SharedKeyDeleteDelay: func() *time.Duration { a := sharedKeyDeleteDelay; return &a }(),
 	})
 	c.Assert(err, IsNil)
 	c.Assert(store, Not(IsNil))

